### ISSUE TITLE
Different output of bn256::G1::hash_to_curve between vanilla & asm branches

### ISSUE THREAD

**Issue by huitseeker** [privacy-scaling-explorations/halo2curves#100]
The following branch: https://github.com/huitseeker/halo2curves/tree/test_asm_output

Implements a test which draws byte strings using a deterministically-seeded RNG, passes them to hash_to_curve, and looks at the representation of the obtained point:
https://github.com/huitseeker/halo2curves/blob/test_asm_output/src/bn256/mod.rs#L28-L58

- It passes when running on the vanilla branch (no features) - without surprise because that's where the expected outputs are drawn from,
    - the command is `cargo nextest run --no-capture -E 'test(test_hash_to_curve_print)'`
- it fails on the 10th and last vector when running with `--features "asm"`.
    - the command is `cargo nextest run --no-capture -E 'test(test_hash_to_curve_print)' --features "asm"`
    
Tested on https://github.com/privacy-scaling-explorations/halo2curves/tree/81a078254518a7a4b7c69fab120621deaace9389 but https://github.com/microsoft/Nova/pull/261 shows this applied to v0.4.0 at least and probably since.

I would suspect a possible missing modular reduction in the field arithmetic somewhere between vanilla / asm, but haven't looked into the cause any deeper.

h/t @srinathsetty @adr1anh

**Comment by jonathanpwang** [comment]
definitely concerning. I did some debug prints, and it seems to be in 
```
*buf = F::from_uniform_bytes(&little);
```
in `hash_to_field`. The input bytes that fail are:
```
[src/hash_to_curve.rs:86] &little = [
    162,
    172,
    219,
    208,
    54,
    164,
    142,
    51,
    4,
    27,
    221,
    171,
    16,
    245,
    57,
    68,
    49,
    66,
    6,
    148,
    171,
    117,
    231,
    94,
    66,
    72,
    13,
    39,
    110,
    253,
    77,
    240,
    181,
    2,
    22,
    166,
    144,
    119,
    69,
    41,
    80,
    11,
    213,
    244,
    12,
    85,
    177,
    54,
    189,
    86,
    89,
    95,
    166,
    140,
    47,
    114,
    28,
    2,
    146,
    131,
    204,
    215,
    179,
    51,
]
```
where "asm" outputs `[src/hash_to_curve.rs:88] buf = 0x0f1b0e9edc375e0da16091aa69e540397260e58209ec94a0dd3ade4ccb62d323` while expected no asm is
`[src/hash_to_curve.rs:88] buf = 0x07ac9d7569d798810c68c4e9a905c2da3d84a8a1386886b2544efca8576d252f`.

**Comment by jonathanpwang** [comment]
This is likely why the `from_raw` function for "asm" unrolls the entire `mac`, etc instructions.

**Comment by davidnevadoc** [comment]
That's right. Unrolling the instructions in `from_u512` as it is done in `from_raw` fixes the test.

**Comment by jonathanpwang** [comment]
@davidnevadoc could you make a PR with this fix? 
I wasn't sure if there was a way to save some code for the unrolling. Maybe the assembly macro just needs to have a non-asm montgomery reduce as well?

**Comment by davidnevadoc** [comment]
I made a draft PR the quick fix (#103). I am a bit concerned about this issue though as I don't quite understand it. 
My first doubt it is: Which of the 2 versions is correct? I am guessing the bug in is the asm version but I'm not 100% sure. I think it would be good to check against test vectors from another lib
.

**Comment by mratsim** [comment]
Normally Montgomery multiplication is correct for inputs in the range [0; 2p).

And Montgomery reduction can work with double sized input in the range [0; pR), R being the Montgomery magic number (1²⁵⁶ in our case), unsure where I found this bound again but I do test it in my library: https://github.com/mratsim/constantine/blob/d77bb79/constantine/math/arithmetic/finite_fields_double_precision.nim#L72-L81

**Comment by jonathanpwang** [comment]
Bump on this issue: is https://github.com/privacy-scaling-explorations/halo2curves/pull/103 an acceptable fix?

I think it's pretty high priority to fix this issue because the `transcript_repr` in halo2 `VerifyingKey` does use `from_uniform_bytes` which uses `from_u512`.

**Comment by mratsim** [comment]
> Normally Montgomery multiplication is correct for inputs in the range [0; 2p).
> 
> And Montgomery reduction can work with double sized input in the range [0; pR), R being the Montgomery magic number (1²⁵⁶ in our case), unsure where I found this bound again but I do test it in my library: https://github.com/mratsim/constantine/blob/d77bb79/constantine/math/arithmetic/finite_fields_double_precision.nim#L72-L81

seems like the assembly code actually mentions the same bound:

https://github.com/privacy-scaling-explorations/halo2curves/blob/91f34666c78e3acaf726030be40240041ea1b8de/src/derive/field.rs#L68-L74

I don't think from u512 is on a perf critical path so we can remove the assembly for now and create an issue to revisit?

**Comment by davidnevadoc** [comment]
I think this is an acceptable temporary fix.

Just to make sure I'm understanding the issue correctly. 
For Bn256 base field `Fq` we have:
```
    q  =  0x30 ...
    r  =  0x0e ...
    r2 =  0x06 ...
    r3 =  0x20 ...
```
Since `r3 > r` we may exceed the Montgomery reduction bound `r * q` for values of `c` when `c * r3 > q *r`. Is this right?

**Comment by davidnevadoc** [comment]
Solved with #108
